pygmt/clib/conversion.py

```diff
@@ -162,19 +162,36 @@ def _to_numpy(data: Any) -> np.ndarray:
         "date64[ms][pyarrow]": np.datetime64,
     }
 
-    if (
-        hasattr(data, "isna")
-        and data.isna().any()
-        and Version(pd.__version__) < Version("2.2")
-    ):
-        # Workaround for dealing with pd.NA with pandas < 2.2.
-        # Bug report at: https://github.com/GenericMappingTools/pygmt/issues/2844
-        # Following SPEC0, pandas 2.1 will be dropped in 2025 Q3, so it's likely
-        # we can remove the workaround in PyGMT v0.17.0.
-        array = np.ascontiguousarray(data.astype(float))
-    else:
-        vec_dtype = str(getattr(data, "dtype", ""))
-        array = np.ascontiguousarray(data, dtype=dtypes.get(vec_dtype))
+    # pandas nullable dtypes and pyarrow types were converted to np.object_ dtype
+    # before, and are converted to suitable numpy dtypes since pandas 2.2.
+    # Refer to the following link for details:
+    # https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#to-numpy-for-numpy-nullable-and-arrow-types-converts-to-suitable-numpy-dtype
+    # Here are the workarounds for pandas < 2.2.
+    # Following SPEC 0, pandas 2.1 should be dropped in 2025 Q3, so it's likely we can
+    # remove the workaround in PyGMT v0.17.0.
+    if Version(pd.__version__) < Version("2.2"):
+        dtypes.update(
+            {
+                "Int8": np.int8,
+                "Int16": np.int16,
+                "Int32": np.int32,
+                "Int64": np.int64,
+                "UInt8": np.uint8,
+                "UInt16": np.uint16,
+                "UInt32": np.uint32,
+                "UInt64": np.uint64,
+                "Float32": np.float32,
+                "Float64": np.float64,
+            }
+        )
+        if hasattr(data, "isna") and data.isna().any():
+            # Integer dtypes with missing values are cast to NumPy float dtypes and NaN
+            # is used as missing value indicator.
+            dtype = np.float64 if data.dtype.kind in "iu" else data.dtype.numpy_dtype
+            data = data.to_numpy(dtype=dtype, na_value=np.nan)
```

Comment:
For pandas objects with missing values, we can't just map `Int32` to `np.float64`, as shown below (with pandas 2.1):
```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: x = pd.Series([1, None, 3], dtype=pd.Int32Dtype())

In [4]: np.ascontiguousarray(x, dtype=np.float64)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.ascontiguousarray(x, dtype=np.float64)

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/series.py:953, in Series.__array__(self, dtype)
    906 """
    907 Return the values as a NumPy array.
    908
   (...)
    950       dtype='datetime64[ns]')
    951 """
    952 values = self._values
--> 953 arr = np.asarray(values, dtype=dtype)
    954 if using_copy_on_write() and astype_is_view(values.dtype, arr.dtype):
    955     arr = arr.view()

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/arrays/masked.py:575, in BaseMaskedArray.__array__(self, dtype)
    570 def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
    571     """
    572     the array interface, return my values
    573     We return an object array here to preserve our scalar values
    574     """
--> 575     return self.to_numpy(dtype=dtype)

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/arrays/masked.py:487, in BaseMaskedArray.to_numpy(self, dtype, copy, na_value)
    481 if self._hasna:
    482     if (
    483         dtype != object
    484         and not is_string_dtype(dtype)
    485         and na_value is libmissing.NA
    486     ):
--> 487         raise ValueError(
    488             f"cannot convert to '{dtype}'-dtype NumPy array "
    489             "with missing values. Specify an appropriate 'na_value' "
    490             "for this dtype."
    491         )
    492     # don't pass copy to astype -> always need a copy since we are mutating
    493     with warnings.catch_warnings():

ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```
So we have to call `.to_numpy()` explicitly.